### ISSUE TITLE
enhancing 11.5.x means dependency on io.ebean:ebean:11.4.1

### ISSUE THREAD

**Issue by tobias-** [ebean-orm/ebean-idea-enhancer#22]
Just a FYI for those (like me) that are a bit confused by the versions. 

io.ebean:ebean-agent:11.5.1 is the dependency the plugin needs, but the dependency in the enhanced code will be against io.ebean:ebean:11.4.1. The plugin does not work with all io.ebean:ebean:11.* as it will fail when starting with something similar to this:

```
FATAL - [Scanner-1] ervletContainerInitializer - Dying because of exception
java.lang.NoSuchMethodError: io.ebean.bean.EntityBeanIntercept.preSetter(ZILjava/lang/Object;Ljava/lang/Object;)V
    at com.example.Redacted._ebean_set_version(Redacted.java:1) ~[core-X.X.XXX-SNAPSHOT.jar:?]
    at com.example.Redacted.<init>(Redacted.java:32) ~[core-X.X.XXX-SNAPSHOT.jar:?]
    at com.example.Redacted2.<init>(Redacted2.java:103) ~[core-9.0.720-SNAPSHOT.jar:?]
    at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_151]
    at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:1.8.0_151]
    at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_151]
    at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_151]
    at java.lang.Class.newInstance(Class.java:442) ~[?:1.8.0_151]
    at io.ebeaninternal.server.deploy.BeanDescriptor.createPrototypeEntityBean(BeanDescriptor.java:577) ~[ebean-11.3.1.jar:?
```

I get the feeling your primary focus is not in the idea plugin (not that I blame you), but maybe it would be relatively easy to make the plugin depend on multiple agents and use the one that is for the current module's dependency? Maybe repackage the different agents (e.g. via the gradle shadow plugin) and call the right one? If it sounds interesting, let me know and I'll spike it.

OT: We're testing the changes I've made to the classpath/classreader part of the plugin to make sure I didn't break anything, but expect a PR within the next day. I've tried and failed to create junit test cases for the plugin, so it's manual testing.

**Comment by rbygrave** [comment]
> I get the feeling your primary focus is not in the idea plugin 

No not at all.  I view it of primary importance !!

> maybe it would be relatively easy to make the plugin depend on multiple agents and use the one that is for the current module's dependency?

Maybe.  I mentioned this on the forum (that we somewhat had a breaking change in the enhancement - https://groups.google.com/forum/#!topic/ebean/vzimy1bhIjk). 

I was initially hoping to be able to support both old and new enhancement 11.x options but it didn't turn out to be easy for me to do so I went with:
- Posting the warning on the forum (for people to take care upgrading the IDEA plugin)
- Pushing forward with the breaking change 

With the view that many in 11.x will bump both agent and ebean and not upgrade the IDEA plugin until they are ready to do that.


> Maybe repackage the different agents

I have my own time pressures like everyone :) ... The question we need to ask is, is it not better/faster to update both Ebean and the agent now.  How long are you going to be using Ebean 11.x less than 11.4 ?

**Comment by tobias-** [comment]
>No not at all. I view it of primary importance !!

Sorry, it was just a feeling I had.

> I was initially hoping to be able to support both old and new enhancement 11.x options but it didn't turn 
out to be easy for me to do so I went with:

No, I actually went ahead and tried shadowing/shading the libraries via the plugin I mentioned, but my forte is not Gradle. I haven't given up, not yet though.

> With the view that many in 11.x will bump both agent and ebean and not upgrade the IDEA plugin until they are ready to do that.

The upgrade button is very tempting every time you start the editor :). That said, you're right. My main problem (and the reason for this issue) was that if I weren't messing around with the plugin, none of the dependencies I use for 11.5-enhancing has 11.5 in the version.


> I have my own time pressures like everyone :)

Yes, and we all appreciate the amount of work you manage to get done.

Btw, subscribed to the forum now, and closing this issue.
